pandas/tests/groupby/test_custom_metadata.py

```diff
@@ -0,0 +1,89 @@
+"""
+Test metadata propagation in groupby
+
+The PandasTable class below is implemented according to the [guidelines],
+and as such would expect `__finalize__` to always be called so that the
+`_pandastable_metadata` is always populated.
+
+[guidelines]: https://pandas.pydata.org/pandas-docs/stable/development/extending.html#override-constructor-properties  # noqa
+"""
+
+from typing import List
+from warnings import warn
+
+import pytest
+
+import pandas as pd
+
+_TABLE_METADATA_FIELD_NAME = "_pandastable_metadata"
+
+
+def _combine_metadata(data: List[str]) -> str:
+    """
+    A mock implementation for testing
+    """
+    return "+".join(data)
+
+
+class PandasTable(pd.DataFrame):
```

Comment:
can this be tested without a subclass?  just set _metadata on a regular dataframe?

Reply:
I don't think so: The `__finalize__` implementation in NDFrame copies all fields from `self._metadata`, but since self is a new object this would do nothing unless `_metadata` has been set for the class. We could make a test using a monkeypatched `DataFrame` class, but that would not catch the error of using `DataFrame` constructor rather than calling `_constructor`.

I could have a go at a more minimal test where only subclass features are `_metadata` and `_constructor` ?

Reply:
Have updated the testcase to use a minimal DataFrame extension. This still catches the original regression.